enhancements/mcd-reboot-exceptions.md

```diff
@@ -0,0 +1,247 @@
+reboots, against their frequency and user impact.
+
+In environments where the limit is generally lower, such as bare metal, the
+balance can be different - creating a need to find alternative ways to apply
+configuration changes for some scenarios.
+
+### Goals
+
+Provide a mechanism for avoiding a full cluster reboot in domain specific
+scenarios where engineering has verified that the configuration change:
+
+1. can be proven to be safely and immediately applied by some other mechanism,
+1. is not combined with other changes that do require a reboot
+1. has a demonstrable negative user impact when applied via reboot
+
+### Non-Goals
+
+1. Removing all reboots from the system
+1. Allowing the admin to configure additional domains or strategies for applying
+   configuration updates
+
+## Proposal
+
+1. Add a function to MCD that compares two MachineConfigs, and outputs an
+   ordered list of known actions that are needed to apply it
+1. Modify any parts of MCD that assume drain and reboot always happens after
+   writing to disk
+1. Modify MCD to perform only the list of actions produced in step 1
+
+The full list of post-write actions required for this feature to be useful are:
+
+- drain
+- reboot
+- restart a named systemd service
+
+If the action results in an error, or takes too long, a reboot will be
+performed as in prior 4.x versions.
+
+### User Stories [optional]
+
+The user stories below are intended to be illustrative of the different types of 
+services we need to build capabilities for.
+
+#### Story 1
+
+As an administrator of a disconnected bare metal cluster, I want the ability to
+update the image catalog mirror[1] without potentially causing a full cluster
+reboot, so that I can obtain the latest fixes quickly and without needing to be
+concerned about the cumulative disruption to the cluster.
+
+[1] As well as the ICSP associated with it, which is sensitive to the addition,
+subtraction, or renaming of images.
+
+#### Story 2
+
+As an adminstrator of a large bare metal cluster, I want the addition or removal
+of SSH keys to happen without a full cluster reboot, so that the credentials are
+usable (or revoked) in a timeframe comparable with traditional environments not
+hours.
+
+#### Story 3
+
+As an adminstrator of a bare metal cluster that takes a day to reboot, I want
+the system to use the least invasive method for applying configuration changes,
+so that I have periods of relative stability where I can accomodate the
+maintenance windows of less enlightened vendors.
+
+#### Story 4
+
+As an admin, I want to continue to force reboots for network configuration 
+updates and changes to kernel flags, so that they are applied in a timely 
+manner.
+
+
+### Implementation Details/Notes/Constraints [optional]
+
+How the comparision determines what constitutes a domain, whether by target file
+or by configuration element, is left for later discussion.
```

Comment:
It is not clear what "domain" means here. It seems to mean something much more specific than in, eg, "Domain specific minimisation of config driven cluster reboots"

Reply:
Reworded to explain the intended usage/meaning